examples/logo_interpreter.rs

```diff
@@ -0,0 +1,75 @@
+//! To run a LOGO program, run:
+//!     cargo run --example logo_interpreter -- .\examples\my_logo_program.txt
+
+extern crate turtle;
+
+use turtle::Turtle;
+use std::env;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::collections::VecDeque;
+
+
+fn main() {
+    let mut turtle = Turtle::new();
+
+    let args: Vec<String> = env::args().collect();
+
+    if args.len() != 2 {
+        println!("No file provided");
+        return
+    }
+
+    let path = &args[1];
+    let file = File::open(path).expect("Could not open provided program");
+    let mut reader = BufReader::new(file);
+    loop {
+        let mut buffer = String::new();
+        let read_bytes = reader.read_line(&mut buffer)
+            .expect("Unable read input from stdin");
+        if read_bytes == 0 {
+            // Reached EOF, stop interpreter
+            break;
+        }
+
+        let mut cmd_args = buffer.split_whitespace().collect();
+        handle_command(&mut turtle, &mut cmd_args);
+    }
+}
+
+fn handle_command(turtle: &mut Turtle, args: &mut VecDeque<&str>) {
```

Comment:
Great idea, it's working properly.

But I don't properly understand `<'a, A: Iterator<Item=&'a str>>`. Can you explain on how you got to this ? 
And if possible share some resources where I can read more about it.

Reply:
Hi @Jay9596. To understand that bit of code, you need to understand a bit about [Generics](https://doc.rust-lang.org/book/second-edition/ch10-01-syntax.html), [Traits and Trait Bounds](https://doc.rust-lang.org/book/second-edition/ch10-02-traits.html) and [Iterators](https://doc.rust-lang.org/book/second-edition/ch13-02-iterators.html). If those links to the Rust book aren't enough, I'm happy to help answer your questions on [Gitter](https://gitter.im/rust-turtle/discuss) or in the comments of this issue.

That might be a lot to understand in only a little bit of time, so let me try to give you a high level explanation that will help you get at least an intuition for what is going on.

The main part of the code in question is as follows:
```rust
fn handle_command<'a, A: Iterator<Item = &'a str>>(turtle: &mut Turtle, mut args: A) {
```
All the code that has to with `'a` is [Rust's lifetimes](https://doc.rust-lang.org/book/second-edition/ch10-03-lifetime-syntax.html) feature in action. That isn't entirely relevant to what we're trying to express, so let's take that out:
```rust
fn handle_command<A: Iterator<Item = &str>>(turtle: &mut Turtle, mut args: A) {
```
Now we have something that is at least a little bit simpler. 

`handle_command` is a function that takes two arguments. Function arguments are our way of expressing what type of value we want our function to accept. We know that we need at least one argument to be the turtle, but what should the second argument be? Let's think about the minimum set of requirements for that type.

In case you aren't familiar with them, an `Iterator` is something that has a `next()` method and can produce some "item" every time `next()` is called. Iterators are often really efficient because they only allocate enough space for the next item and never the entire set of items that they can produce. When we were calling `collect()` earlier, we were forcing `split_whitespace()` (an Iterator) to allocate space for all of its items when we really didn't need it to.

In reality, our `handle_command` function is just going to look at each argument in the buffer one at a time. If it finds an error, it'll panic and make the rest of the buffer irrelevant anyway. Using an `Iterator` allows us to express that we want **anything** that meets these properties. When we say `Iterator<Item = &str>`, we're expressing that we want the item produced by the iterator to be a string.

So why bother with all the funky syntax like `A: Iterator<Item = &str>`? Well it allows us to say "we want some type that produces strings when we call `next()`" without actually specifying a type that does that. That means that if we decide to use something other than `split_whitespace()` later, we won't have to update `handle_command()` to do so. As long as we pass *something* that produces `&str` when we call `next()`, Rust will be happy.

There's much more explanation about the details of all of that in the chapters I linked you to above. Once again, please don't hesitate to ask me questions. These topics might seem a bit complicated, but if you take some time to digest them and ask questions, I'm sure you'll be able to figure them out. :smile:

Reply:
The high level explanation was all that I needed, I just wanted to know why should we use iterator and what the funny syntax `A: Iterator<Item = &str>` did.

It basically tells the Rust compiler that `A` a generic type implements an iterator that will produce some `&str`. So now we can even pass a user defined struct as long as it implements `Iterator` and produces `&str` when iterated without having to change a single thing in `handle_command()`.

Thank you for explaining it properly. 👍

Reply:
Yes! You've got it! :D